File: scala/src/main/scala/adt/command.scala
package adt

enum Direction {
  case North, East, South, West
}

final abstract class Idle
final abstract class Moving

sealed trait Command[Before, After]

object Command {
  case class  Face(dir: Direction) extends Command[Idle, Idle]
  case object Start  extends Command[Idle, Moving]
  case object Stop   extends Command[Moving, Idle]

  case class Chain[A, B, C](
      cmd1: Command[A, B],
      cmd2: Command[B, C]
  ) extends Command[A, C]
}

implicit class Compose[A, B](cmd1: Command[A, B]) {
  def ->[C](cmd2 : Command[B, C]) : Command[A,C] = Command.Chain(cmd1, cmd2)
}

@main
def main() = {
  val c1 : Command[Idle, Idle] = Command.Chain(Command.Chain(Command.Face(Direction.North), Command.Start), Command.Stop)
  val c2 = Command.Face(Direction.North) -> Command.Start -> Command.Stop
 // val c3 = Command.Start -> Command.Face(Direction.North)
 // val c4 = Command.Chain(Command.Start,  Command.Face(Direction.North))
 if (c1.equals(c2)) {
   println("OK")
 }
}

File: java/src/main/java/org/example/Command.java
package org.example;

import java.util.ArrayList;
import java.util.List;

public sealed interface Command<Before extends Command.State, After extends Command.State>
    permits Command.Face, Command.Start, Command.Stop, Command.Chain, Command.Composite {

  sealed interface State permits Command.Idle,  Command.Moving {}

  final class Idle implements State {}
  final class Moving implements State {}

  default <C extends Command.State> Command<Before, C> add(Command<After, C> cmd) {
    return new Composite<Before, Before>().add(this).add(cmd);
  }

  record Face(Direction direction) implements Command<Idle, Idle> {
  }

  final class Start implements Command<Idle, Moving> {
    private Start() {}
    private static final Start instance = new Start();
    public static Start getInstance() {
      return Start.instance;
    }
  }

  final class Stop implements Command<Moving, Idle> {
    private Stop() {}
    private static final Stop instance = new Stop();
    public static Stop getInstance() {
      return Stop.instance;
    }
  }

  final class Chain<A extends Command.State, B  extends Command.State, C  extends Command.State> implements Command<A, C> {
    private final Command<A, B> cmd1;
    private final Command<B, C> cmd2;

    private Chain(final Command<A, B> cmd1, final Command<B, C> cmd2) {
      this.cmd1 = cmd1;
      this.cmd2 = cmd2;
    }
  }

  final class Composite<A extends Command.State, B  extends Command.State>  implements Command<A, B> {
    private List<Command> commands = new ArrayList<>();

    @Override
    public <C extends Command.State> Command<A, C> add(Command<B, C> subCommand) {
      if (subCommand instanceof Command.Composite<B,C> cp) {
        this.commands.addAll(cp.commands);
      }
      else {
        this.commands.add(subCommand);
      }
      return (Command<A, C>) this;
    }
  }

  static <A extends Command.State, B  extends Command.State, C  extends Command.State> Command<A, C> build(Command<A, B> c1,
      Command<B, C> c2) {
    return new Chain<>(c1, c2);
  }


  enum Direction {
    NORTH, EAST, SOUTH, WEST;
  }
}
